uniffi_bindgen/src/bindings/gecko/templates/SharedHeaderTemplate.h

```diff
@@ -0,0 +1,795 @@
+    MOZ_RELEASE_ASSERT(size.isValid());
+    return size.value();
+  }
+
+  static bool ReadFrom(Reader& aReader, nsACString& aValue) {
+    aValue.Truncate();
+    aReader.ReadRawString<nsACString>(
+        aValue, [](Span<const char> aRawString, nsACString& aValue) {
+          aValue.Append(aRawString);
+        });
+    return true;
+  }
+
+  static void WriteInto(Writer& aWriter, const nsACString& aValue) {
+    aWriter.WriteRawString(aValue.Length(), [&](Span<char> aRawString) {
+      memcpy(aRawString.Elements(), aValue.BeginReading(), aRawString.Length());
+      return aRawString.Length();
+    });
+  }
+};
+
+template <>
+struct ViaFfi<nsACString, char*> {
+  static bool Lift(const char*& aLowered, nsACString& aLifted) {
+    aLifted.Truncate();
+    aLifted.Append(MakeStringSpan(aLowered));
+    return true;
+  }
+
+  static char* Lower(const nsACString& aLifted) {
+    RustError err = {0, nullptr};
+    char* lowered = {{ ci.ffi_string_alloc_from().name() }}(aLifted.BeginReading(), &err);
+    if (err.mCode) {
+      MOZ_ASSERT(false, "Failed to copy narrow string to Rust string");
+      return nullptr;
+    }
+    return lowered;
+  }
+};
+
+/// Shared traits for serializing `nsString`s and `nsAString`s.
+template <typename T>
+struct StringTraits {
+  static size_t Size(const T& aValue) {
+    auto size = EstimateUTF8Length(aValue);
+    size += sizeof(uint32_t);  // For the length prefix.
+    MOZ_RELEASE_ASSERT(size.isValid());
+    return size.value();
+  }
+
+  static bool ReadFrom(Reader& aReader, T& aValue) {
+    aValue.Truncate();
+    aReader.ReadRawString<T>(aValue,
+                             [](Span<const char> aRawString, T& aValue) {
+                               AppendUTF8toUTF16(aRawString, aValue);
+                             });
+    return true;
+  }
+
+  static void WriteInto(Writer& aWriter, const T& aValue) {
+    auto length = EstimateUTF8Length(aValue);
+    MOZ_RELEASE_ASSERT(length.isValid());
+    aWriter.WriteRawString(length.value(), [&](Span<char> aRawString) {
+      return ConvertUtf16toUtf8(aValue, aRawString);
+    });
+  }
+
+  /// Estimates the UTF-8 encoded length of a UTF-16 string. This is a
+  /// worst-case estimate.
+  static CheckedInt<size_t> EstimateUTF8Length(const T& aUTF16) {
+    CheckedInt<size_t> length(aUTF16.Length());
+    // `ConvertUtf16toUtf8` expects the destination to have at least three times
+    // as much space as the source string.
+    length *= 3;
+    return length;
+  }
+};
+
+template <>
+struct Serializable<nsAString> {
+  static size_t Size(const nsAString& aValue) {
+    return StringTraits<nsAString>::Size(aValue);
+  }
+
+  static bool ReadFrom(Reader& aReader, nsAString& aValue) {
+    return StringTraits<nsAString>::ReadFrom(aReader, aValue);
+  }
+
+  static void WriteInto(Writer& aWriter, const nsAString& aValue) {
+    StringTraits<nsAString>::WriteInto(aWriter, aValue);
+  }
+};
+
+template <>
+struct ViaFfi<nsAString, char*> {
```

Comment:
Technically, we can specialize `ViaFfi` for `nsString`, too—I didn’t because we only lift `char *`s into `nsAString&` out params, not `nsString`s on the heap or `nsAutoString`s on the stack.